### DISCUSSION TITLE
Settle on an approach for providing additional features

### DISCUSSION THREAD

**Discussion by stdstaff** [govuk-ruby/govuk-design-system-rails#83] [Ideas]
As part of the team's [community readiness work](https://github.com/orgs/govuk-ruby/projects/2) we are seeking to make the gem compliant with the [GDS contribution criteria](https://design-system.service.gov.uk/community/contribution-criteria/#developing-a-community-resource-or-tool). This means being feature complete and generating the same HTML output as `govuk-frontend`.

However, our gem also provides some enhancements and extra features which are not found within the Design System. For example, we include some [HMCTS Components](https://github.com/govuk-ruby/govuk-design-system-rails/wiki/Components#hmcts-frontend), and an [Accessible Autocomplete](https://github.com/govuk-ruby/govuk-design-system-rails#component-implementations) enhancement to `govukSelect`.

The upstream [HMCTS Frontend](https://github.com/hmcts/frontend) project has been archived and we will likely remove these components in a future release. However, the Accessible Autocomplete we have added is in use by a number of government services.

Having [the autocomplete as part of this project](https://github.com/govuk-ruby/govuk-design-system-rails/blob/6fe297831a6cc7b8adebe91962a5786ad04bab41/app/views/components/_govuk_select.html.erb#L87-L106) is problematic: It means the gem can generate HTML which is not part of the expected GDS output, and, that functionality relies on [JavaScript which is not bundled with the project](https://github.com/OfficeForProductSafetyAndStandards/product-safety-database/blob/8a3763c42ba30c82a32a55f7cc7ff034652c2540/app/javascript/autocomplete.js). Host applications are expected to include this JavaScript to use this feature, but it is never mentioned anywhere in our README and was never extracted from the original service.

Additionally, contributors to the project occasionally [propose enhancements to the GDS components](https://github.com/govuk-ruby/govuk-design-system-rails/pull/79) which may be useful within their services or within the context of a Rails application but which do not form part of the Design System or `govuk-frontend`. We cannot merge these without impacting the tenets of the GDS criteria.

Please discuss approaches we can take clean up `govuk-design-system-rails`.

Suggestions:
 - Remove the HMCTS Frontend components
 - Create a `govuk-design-system-rails-extensions` gem and move the autocomplete functionality there
 - Bundle the javascript required for autocomplete with the new gem
 - Add contributor guidelines to this project asking anyone who is adding functionality to first propose it in `govuk-frontend`/to the GDS team, and otherwise, to propose it in `govuk-design-system-rails-extensions`
 - Try to find any other inconsistencies which already exist (for example [this PR](https://github.com/govuk-ruby/govuk-design-system-rails/pull/61), while useful, looks like an extension) and move them to the `-extensions` gem

**Comment by slorek** [comment]
I think it's fair to remove the HMCTS components and move them to another gem. I think it makes sense to be explicit that these are HMCTS components and name that gem accordingly. Then downstream users can include them only when needed.

The accessible autocomplete gem is maintained separately from the design system and I think a similar approach could be used for that. We could consider naming the helpers uniquely so that usage is explict rather than overriding the original implementation.

I suspect bundling the Javascript may be difficult due to the differing asset pipeline options out there and the fact the Node package is separate to the Ruby gem dependencies. I think including instructions in  in the `README` should be sufficient.
